syntax = "proto3";
package train_conductor.service.training;
import "google/protobuf/timestamp.proto";

/*-- SERVICES ----------------------------------------------------------------*/

service TrainConductor {
  rpc Train(TrainingRequest) returns (TrainingJob);
  rpc GetTrainingStatus(TrainingInfoRequest) returns (TrainingStatusResponse);
  rpc CancelTraining(TrainingInfoRequest) returns (TrainingStatusResponse);
}

/*-- MESSAGES ----------------------------------------------------------------*/

message TrainingJob {

  /*-- fields --*/
  string training_id = 1;
  string model_name = 2;
}

message TrainingRequest {

  /*-- fields --*/
  string model_name = 1;
  string output_path = 2;
  TrainingParameters parameters = 3;
}


message TrainingParameters {

  /*-- fields --*/
  string base_model = 1;
  string data_path = 2;
  optional bool bf16 = 3;
  optional int64 num_train_epochs = 4;
  optional int64 per_device_train_batch_size = 5;
  optional int64 per_device_eval_batch_size = 6;
  optional int64 gradient_accumulation_steps = 7;
  optional string evaluation_strategy = 8;
  optional string save_strategy = 9;
  optional double learning_rate = 10;
  optional double weight_decay = 11;
  optional double warmup_ratio = 12;
  optional string lr_scheduler_type = 13;
  optional int64 logging_steps = 14;
  optional string fsdp = 15;
  optional string fsdp_config = 16;
  optional bool include_tokens_per_second  = 17;
  optional bool packing = 18; 
  optional string response_template = 19;
  optional string dataset_text_field = 20;
}


message TrainingInfoRequest {
  string training_id = 1;
}

message TrainingStatusResponse {

  /*-- fields --*/
  string training_id = 1;
  TrainingStatus state = 2;
  google.protobuf.Timestamp submission_timestamp = 3;
  google.protobuf.Timestamp completion_timestamp = 4;
  repeated string reasons = 5;
}

/*-- ENUMS -------------------------------------------------------------------*/

enum TrainingStatus {
  PLACEHOLDER_UNSET = 0;
  PENDING = 1;
  QUEUED = 2;
  RUNNING = 3;
  SUSPENDED = 4;
  COMPLETED = 5;
  CANCELED = 6;
  FAILED = 7;
  DELETED = 8;
}



